README.md
# Fullcalendar Plugin

The **Fullcalendar** Plugin is for [Grav CMS](http://github.com/getgrav/grav). It reads ICS Calendar Files and shows Events in a monthly Calendar Widget on your Page(s) - including Month Specific Images (if available in the calendar page folder):

![](monthpic.png)

## Installation

Installing the Fullcalendar plugin can be done in one of two ways. The GPM (Grav Package Manager) installation method enables you to quickly and easily install the plugin with a simple terminal command, while the manual method enables you to do so via a zip file.  
### GPM Installation (Preferred)

The simplest way to install this plugin is via the [Grav Package Manager (GPM)](http://learn.getgrav.org/advanced/grav-gpm) through your system's terminal (also called the command line).  From the root of your Grav install type:

    bin/gpm install fullcalendar

This will install the Fullcalendar plugin into your `/user/plugins` directory within Grav. Its files can be found under `/your/site/grav/user/plugins/fullcalendar`.  
### Manual Installation

To install this plugin, just download the zip version of this repository and unzip it under `/your/site/grav/user/plugins`. Then, rename the folder to `fullcalendar`. You can find these files on [GitHub](https://github.com/wernerjoss/grav-plugin-fullcalendar) or via [GetGrav.org](https://getgrav.org/downloads/plugins).  
You should now have all the plugin files under

    /your/site/grav/user/plugins/fullcalendar

> NOTE: This plugin is a modular component for [Grav](http://github.com/getgrav/grav) and requires the [Error](https://github.com/getgrav/grav-plugin-error), [Problems](https://github.com/getgrav/grav-plugin-problems) and [Shortcode Core](https://github.com/getgrav/grav-plugin-shortcode-core) Plugins to operate.  
### Admin Plugin

If you use the admin plugin, you can install directly through the admin plugin by browsing the `Plugins` tab and clicking on the `Add` button.  
## Configuration

Before configuring this plugin, you should copy the `user/plugins/fullcalendar/fullcalendar.yaml` to `user/config/plugins/fullcalendar.yaml` and only edit that copy.  
Note that if you use the admin plugin, the file with your configuration, named fullcalendar.yaml will be saved in the `user/config/plugins/` folder once the configuration is saved in the admin.  
There is also a simple UI in the admin Backend that provides basic customizations.  
Here is the default configuration and an explanation of available options:

```yaml

enabled: true
colors: "#3a87ad" # see additional Note on custom colors in the Changelog, be sure to use single or double quotes around list !
showlegend: false   # set to true to show calendar File Name(s) as Legend below grid
weekNumbers: false  # set to true to show Week Numbers
cors_api_url:   # leave this empty if you are not sure - included local CORS proxy will be evaluated automatically !
enableOnAllPages: false    # Formerly known as useCustomPageTemplate, a name that was not clear what it is really used for - better use page Frontmatter (see usage)
```

## Usage

Once installed and enabled, you can use this Plugin to parse ICS Calendar File(s) - these must be found in user/data/calendars e.g. like this:
```
user/data/calendars
├── events.ics
└── holidays.ics
```
and set as parameter in the Plugin shortcode (without Path !) and display Events from that Calendar(s) anywhere on your Site adding this:

    [fullcalendar icsfile="events.ics,holidays.ics,..."][/fullcalendar]

to the content of the appropriate page (note the double quotes " surrounding the file name - single quotes ' will not work !)  
This is the standard method, however, from Version 0.2.6, it is also possible to just drop .ics Calendar Files into your page folder:
```
user/pages
├── 01.home
├── 02.typography
└── 03.events
    ├── calendar.md
    ├── events.ics
    └── holidays.ics
```
These will be picked up automatically and used like those in user/data/calendars.  
In case you only use calendar files in the page folder, be sure to include an empty shortcut:  
` [fullcalendar][/fullcalendar]`
in your page content, otherwise it will not work !  
Instead of local Files, you can also provide absolute URL's to (remote) ICS Calendars, in which case a CORS proxy will be used to access them - see [this Document](ExternalCalendars.md) on how to use external Calendars.  
Also note, that from v 0.2.8, the Plugin is supposed to use the calendar.html.twig template from the plugin (or a modified copy in your Theme Folder) for the calendar page - this can be done manually or in the admin backend by choosing 'Calendar' in the dropdown for the page template.  
From v0.3.0, the Plugin can also be enabled on a per-page base via frontmatter: just set `fullcalendar: true` in the page header.  
This Option is useful in case the calendar.html.twig template cannot be used, e.g. for a calendar in a modular page or in a blog.  
Alternatively, there is still the Configuration Option useCustomPageTemplate (this Name is obsolete, as it led to confusion, new Name is now enableOnAllPages), which, if set to true, will enable the Plugin globally on each and every page.  
However, this is not recommendend, due to the heavy usage of javascript Librarys from fullcalendar.io - these will affect the whole site.  
For that reason, the useCustomPageTemplate/enableOnAllPages Flag is no more available in the Admin Backend (but still working if set manually in the config file).  
In short, the use of the Default page template is recommended for most use cases, as only this way, you will be able to use the monthly picture feature or the handy calendar page drop-in for ics files.  
As mentioned, you can also show a Picture for the current month above the calendar widget (see screenshot obove), if using the standard template.  
Just put 12 Image Files named 'January.jpg', 'February.jpg', ... , 'December.jpg' in the Folder for your Page where the Calendar will be placed.  
(Note that Image File names must match Month names according to your locale setting, so, for locale: de, use 'Januar.jpg' ...).  
Finally, it should be noted that this Plugin relies on jquery, which is loaded by most Themes - in case you are using a Theme that does not do this, jquery is now (from v 0.2.12) automatically loaded, no more config Option needed for that.  
## Advanced Usage
As an addition to the standard use case, there is an elegant way to automatically update your .ics Files from remote Calendars in case those are hosted on a CalDav Server (e.g. Owncloud, Nextcloud...):
In this case, you can just use [caldav2ics](https://github.com/wernerjoss/caldav2ics) via cron job or the [Grav Scheduler](https://learn.getgrav.org/17/advanced/scheduler) to automatically update your ics Files shown by the Fullcalendar Plugin, so that remote Calendar content, usually maintained in separate Calendar Apps (such as Google Calendar or Lightning) is automatically propagated to your Website.  
The same is true if you install [grav-plugin-caldav2ics](https://github.com/wernerjoss/grav-plugin-caldav2ics) which is fully integrated in Grav and has a nice Admin Backend for easy configuration (this will create a config File that can directly be used by caldav2ics).  
## CORS Issues:
As of Begin 2021, I realized, that the recommended external CORS Proxy (herokuapp) will no longer work for public use, see [this support Thread](https://github.com/Rob--W/cors-anywhere/issues/301).  
So I decided to implement a local CORS Proxy, which is available from v 0.2.8.  
For most use cases, the empty cors_api_url in the settings should be ok, as the included internal Proxy will be used, if required.  
Only in case you really prefer to use an external CORS Proxy, you will need to adapt this.  
Hint: the automatically evaluated CORS Proxy URL (created when cors_api_url is empty - so, when updating, be sure to delete the old herokuapp URL if present) is normally http://yourdomain/user/plugins/fullcalendar/proxy.php/ but will be evaluated automatically even if Grav is e.g. installed in Subdirectory grav (or else) !
## Timezone Issues:
This refers to [issue #44](https://github.com/wernerjoss/grav-plugin-fullcalendar/issues/44) which has recently come up and turned out to be a major mess with fullcalendar.io v4 (referred to fc4 in the following).  
Although having added the timezone-aware js Librarys from the fc4 packages, I could not find a way to automatically adjust for the diverse practical use cases/calendar incarnations.  
So I decided to implement some additional settings for the plugin, which can be used to correct for date/time shifts for single/recurring events (yes, it is unfortunately necessary to handle those separately, even w.r. to Daylight Saving Times).  
Anyone having issues with incorrectly displayed Events Times, can now fiddle with these settings until everything looks ok - see the Hints in the admin backend.  
Of course, this is not a really satisfactory Solution, I think the real cause is buried in some bugs of the fc4 release. Maybe (hopefully) the fc5 incarnations are better suited to this - porting this Plugin to fc5 is planned anyway, but will probably not happen in the near future.  
So, for the time beeing, I hope most users can live with the workarounds presented here from Plugin Version 0.3.0.

## Credits

This Plugin is built on [fullcalendar.io](https://fullcalendar.io), [jakubroztocil/rrule](https://github.com/jakubroztocil/rrule) and [jsical](http://mozilla-comm.github.io/ical.js) - Javascript parser for rfc5545

## To Do

* Implement EXDATE/EXRULE rrule exceptions
* Upgrade included Fullcalendar.io to Version 5 (currently: Version 4)


CHANGELOG.md
# v0.3.5
##  07/18/2023

1. [](#new)
    * fix template Selection from Grav Admin

# v0.3.4
##  03/20/2023

1. [](#new)
    * make first Weekday configurable

# v0.3.3
##  09/01/2022

1. [](#bugfix)
    * fix issue #53, google-ics german Umlaut not converted

# v0.3.2
##  04/15/2022

1. [](#new)
    * add minutes Timezone Offset config Option
2. [](#bugfix)
    * fix day shift for recurring all-day events
3. [](#new)
    * merge docs improvement pull requests #48 + #49 from elgandoz
3. [](#bugfix)
    * fix issue #50, calendar file not found for non-root calendar page

# v0.3.0
##  01/10/2022

1. [](#bugfix)
    * fix issue #44 (Incorrect time shown for recurring items)
2. [](#new)
    * improve external calendar usage documentation
    * provide more default values in fullcalendar.yaml
    * add Event Description Popup Option
    * add Option to enable Plugin selective via page frontmatter (instead of global)

# v0.2.13
##  11/11/2021

1. [](#bugfix)
    * fix issue #38 ($config undefined when using custom page template)

# v0.2.12
##  11/07/2021

1. [](#new)
    * merged pull request #36 from @aloxe:
    * remove the need to add jQuery if not done by Theme
    * find correct URL for proxy.php even when Grav isn't installed at the root
    * respect user provided CORS Url if present
    * fix Legend display toggle
    * fix error when Language not set in Grav
2. [](#new)
    * merged pull request #37 from @aloxe

# v0.2.11
##  10/31/2021

1. [](#new)
    * Added Color support from ics files (merge pull request #33 from Johannes Loose)
    * Add config Option useCustomPageTemplate - useful e.g. for modular Calendar Pages (caution: affects page speed if set to true !)

# v0.2.10
##  06/14/2021

1. [](#new)
    * add config Option to add jquery asset in case the used Theme does not.
    * add Frontmatter Variable lang to allow override of language for page

# v0.2.9
##  05/20/2021

1. [](#new)
    * use grav.language.getLanguage instead of config Options - Hint from Ulrich Wolf, thanks :-)

# v0.2.8
##  05/12/2021

1. [](#new)
    * add local CORS Proxy (as default external proxy herokuapp stopped working)

# v0.2.7
##  02/26/2021

1. [](#bugfix)
    * fix weekNumbers setting not recognized, code formatting

# v0.2.6
##  02/20/2021

1. [](#new)
    * improve Monthpics search/display (disable javascript month change event watching when no month pics available)
    * introduce page Template in Plugin, make this selectable in Admin Backend without copy to Theme Templates Folder
    * move javascript out of twig template into separate js File for easier Maintenance
    * enhance blueprints.yaml for easier configuration in admin backend

# v0.2.5
##  09/05/2020

1. [](#new)
    * fix missing event time display when allDay=false
    * fix picture artifact in html when no month pics in page folder
    * add experimental config option tz_offset as a workaround for wrong event times in repeating events (see source code..)

# v0.2.4
##  07/21/2020

1. [](#new)
    * add config Option cors_api_url (overrides default value, if present)

# v0.2.3
##  06/22/2020

1. [](#new)
    * add Event Description as Tooltip

# v0.2.2
##  04/06/2020

1. [](#new)
    * automatically add CORS proxy URL for remote ics Files

# v0.2.1
##  02/17/2020

1. [](#new)
    * fix parsing long lines in ics File(s)
    * new config Option weekNumbers (show week Numbers in calendar grid)
    * fix BYDAY n+- Bug (bysetpos)

# v0.2.0
##  01/06/2020

1. [](#new)
    * update fullcalendar.io to V 4.3.1, including rrule Plugin
    * rewrite ICAL to fullcalendar.io Objects Mapping to support repeating Events (rrule)
    * new locale configuration Option

Repeating Events are now basically supported by incorporating the rrule Plugin from fullcalendar.io V 4.3.1.
There are, however, some rrule Options left, that are not yet supported - should work for most common rrules, though.
Additionaly, a new configuration Option has been introduced: locale - just set this to your Preferred value in fullcalendar.yaml.
Default is en.
Also note that file assets/custom.css has been renamed to daygrid.css - custom.css is for user adaptions and will also be used, if present (but is not provided with plugin) !

# v0.1.5
##  12/30/2019

1. [](#new)
    * allow custom colors for multiple calendars

Calendars can now be assigned custom colors for nicer display by adding an entry 'colors' to fullcalendar.yaml - this should be a comma-separated list of HTML
colors (predifined Names like lightBlue can be used as well as HEX notation, like #ffaabc).
If colors entry is not defined, standard fullcalendar.css color is used for all calendars, same behaviour as before.
In Addition, a Legend (html List, css class cal_legend) can be enabled by adding showlegend: true

# v0.1.4
##  07/30/2019

1. [](#new)
    * allow remote ics file URL's in shortcode, fix automatic locale evaluation

Calendar files can now also be remote, on other servers, these can also be mixed with local calendar files.
In this case, the complete absolute URL to the file(s) has to be provided in the shortcode, as well as enclosing quote signs (") around the whole icsfile= specification, like

    [fullcalendar icsfile="https://calendar.google.com/calendar/ical/myname%40myorganization.com/public/basic.ics" ][/fullcalendar]


assets/calendar.js
// this is now a standalone javascript file, formerly embedded in fullcalendar.html.twig (ugly)
// gets Parameters via DOM, see below

// Load jQuery when it is not loaded already by the theme
if (typeof jQuery=='undefined') {
	var headTag = document.getElementsByTagName("head")[0];
	var jqTag = document.createElement('script');
	jqTag.type = 'text/javascript';
	jqTag.src = 'https://cdnjs.cloudflare.com/ajax/libs/jquery/3.6.0/jquery.min.js';
	jqTag.onload = whenJqReady;
	headTag.appendChild(jqTag);
} else {
	whenJqReady();
}

function whenJqReady() {
	var verbose = false;
	var defaultLocale = 'en';
	var cfgWeekNums = jQuery('#weeknums').text();	//	get Paramter from DOM
	weekNums = false;
	var firstWeekDay = 0;	// Default, new 20.03.23
	var cfgfirstWeekDay = jQuery('#1stweekday').text();	//	get Paramter from DOM
	if (Number(cfgfirstWeekDay) > 0)	firstWeekDay = Number(cfgfirstWeekDay);
	if (verbose)	console.log("1st Weekday: ", firstWeekDay);
	if (cfgWeekNums > 0)	weekNums = true;
	if (verbose)	console.log('Weeknums:', weekNums);
	var cfgLocale = jQuery('#cfgLocale').text();	//	get Paramter from DOM
	var LocaleCode = (cfgLocale !== null) ? cfgLocale : defaultLocale;
	if (verbose)	console.log('LocaleCode:', LocaleCode);

	var pageFilestring = jQuery('#pagecalendars').text();	//	get Paramter from DOM
	if (verbose) console.log('pagecalendars:', pageFilestring);
	var pagecalendars = [];
	if (pageFilestring) {
		pagecalendars = JSON.parse(pageFilestring);
	}
	var calUrls = [];
	var calNames = [];
	var loc = window.location;  // the current page
	pagecalendars.forEach(function(value, index) {
		if (value) {
			url = loc + '/' + value;
			calUrls.push(url);
			calNames.push(value);
		}
	})

	if (verbose)	console.log('pagecalendars:', pagecalendars);

	var cfgFilestring = jQuery('#cfgFilestring').text();	//	get Paramter from DOM
	if (verbose) console.log('cfgfilestring:', cfgFilestring);
	// split string into multiple ics files, if appropriate, see note above
	var cfgfiles = cfgFilestring.split(",").map(function(item) {
	  return item.trim();
	});
	if (verbose) console.log('cfgfiles[]:', cfgfiles);
	var BgColstring = jQuery('#BgColstring').text();	//	get Paramter from DOM'
	if (verbose) console.log('BgColstring:', BgColstring);

	// automatically detect CORS api url from Absolute Path
	var BaseUrl = jQuery('#BaseUrl').text();	//	06.11.21   get Parameter from DOM
	var cors_api_url = BaseUrl + '/user/plugins/fullcalendar/proxy.php/';
	if (verbose) console.log('CORS API URL:', cors_api_url);

	var cfg_cors_api_url = jQuery('#CorsUrl').text();	//	get Paramter from DOM
	if (verbose)	console.log(cfg_cors_api_url.length);
	if (cfg_cors_api_url.length > 8)	{	// use this if it has a reasonable lenth
		if (!cfg_cors_api_url.endsWith('/')) cfg_cors_api_url = cfg_cors_api_url + '/'; // add trailing slash if not present
		cors_api_url = cfg_cors_api_url;
	}

	if (verbose) console.log('CORS Url:', cors_api_url);
	var cfgUrls = [];
	cfgfiles.forEach(function(value, index) {
		cfgFile = value;
		if (value){
			var calName = value;
			if (verbose) console.log('yaml CFG File:' + cfgFile);
			// allow remote ics files, these are handeled by local CORS proxy now
			if (cfgFile.startsWith("https://") || cfgFile.startsWith("http://")) {	// calendar URL is remote
				// automatically add CORS proxy URL for remote calendars, if not yet done 06.04.20 - this is obsolete from v 0.2.8 - local proxy
				calendarUrl = cfgFile;	// always :-)	-	see axjax proxy below 12.05.21
				var index = cfgFile.lastIndexOf("/") + 1;
				calName = cfgFile.substr(index);	// calName is used for Legend, should be only Name, NOT full Url !
			}   else	{
				//	calendarUrl = getAbsolutePath() + 'user/data/calendars/' + cfgFile;	// does not work for non-root calendar pages, see issue #50
				calendarUrl = BaseUrl + '/user/data/calendars/' + cfgFile;	// 15.04.22
			}
			if (verbose) console.log('Calendar URL:' + calendarUrl);
			cfgUrls.push(calendarUrl);
			calNames.push(calName);
		}
	})
	jQuery.merge(calUrls, cfgUrls);
	var len = calUrls.length;
	if (verbose)	console.log('len:', len);
	var colors = BgColstring.split(',');
	var ncolors = colors.length;
	if (ncolors < len)	{	// populate colors with default color
		for (i=ncolors; i<(len);i++)
			colors.push('#3a87ad');
	}
	if (verbose) console.log('colors[]:', colors);
	var showlegend = jQuery('#showlegend').text();	//	get Paramter from DOM'
	showlegend = (showlegend > 0) ? showlegend : false;
	if (verbose)	console.log('showlegend:', showlegend);

	var defaultEnableDescPopup = false;	// DONE: make this configurable via admin
	var cfg_enableDescPopup = jQuery('#enableDescPopup').text();
	var enableDescPopup = (cfg_enableDescPopup !== null) ? cfg_enableDescPopup : defaultEnableDescPopup;

	var cfg_tz_offset_single = jQuery('#tzoffset_single').text();	//	Offset for single Events
	var default_tz_offset_single = 0;	// Default
	var tz_offset_single = (cfg_tz_offset_single !== null) ? cfg_tz_offset_single : default_tz_offset_single;

	var cfg_tz_offset_minutes = jQuery('#tzoffset_minutes').text();	//	minutes Offset for single Events
	var default_tz_offset_minutes = 0;	// Default
	var tz_offset_minutes = (cfg_tz_offset_minutes !== null) ? cfg_tz_offset_minutes : default_tz_offset_minutes;

	var cfg_tz_offset_rec = jQuery('#tzoffset_recur').text();	//	Offset for single Events
	var default_tz_offset_rec = 0;	// Default
	var tz_offset_rec = (cfg_tz_offset_rec !== null) ? cfg_tz_offset_rec : default_tz_offset_rec;

	var useIcsTimezone = jQuery('#useIcsTimezone').text();	//	get Paramter from DOM'
	useIcsTimezone = (useIcsTimezone > 0) ? useIcsTimezone : false;
	var icsTimezone = 'Europe/Berlin';	// Default
	var dlstart = 0;
	var dlend = 0;
	// determine Time Zone from 1st Calendar in List - others will be ignored because time zone is a single setting for the whole calendar Object !
	if (len > 0) {
		calendarUrl = calUrls[0];
		//	console.log(calendarUrl);
		jQuery.ajax({
			crossOrigin: true,
			proxy: cors_api_url,
			url: calendarUrl,
			async: false,	// important !
			context: {},
			success: function(data) {
			}
		})
		.done(function( data, textStatus, jqXHR ) {
			//	console.log(data);
			var jcalData = ICAL.parse(data);	//	directly parse data, no need to split to lines first ! 14.02.20
			var comp = new ICAL.Component(jcalData);

			var tzComps = comp.getAllSubcomponents("vtimezone");
			tzids = jQuery.map(tzComps, function(item) {
				var entry = item.getFirstPropertyValue("tzid");
				if (entry !== null)	icsTimezone = entry;
				// now evaluate Daylight saving time period (months):
				var dlComps = item.getAllSubcomponents("daylight");
					dlids = jQuery.map(dlComps, function(item) {
					var entry = item.getFirstPropertyValue("rrule");
					if (entry !== null)	{
						var parts = entry["parts"];
						//	console.log(parts);
						var bymo = parts["BYMONTH"];	// uppercase !!
						//	console.log(bymo);
						dlstart = parseInt(bymo[0]);
					}
				});
				dlComps = item.getAllSubcomponents("standard");
					dlids = jQuery.map(dlComps, function(item) {
					var entry = item.getFirstPropertyValue("rrule");
					if (entry !== null)	{
						var parts = entry["parts"];
						//	console.log(parts);
						var bymo = parts["BYMONTH"];	// uppercase !!
						//	console.log(bymo);
						dlend = parseInt(bymo[0]);
					}
				});
			});
			if (dlstart > dlend) {	// swap dlstart, dlend if necessary
				temp = dlstart;
				dlstart = dlend;
				dlend = temp;
			}
			//	console.log("daylight start:", dlstart);	console.log("daylight end:", dlend);
		},'text');
	}
	if (verbose)	console.log('icsTimezone: ', icsTimezone);
	var calTimezone = 'local';	// Default
	if (useIcsTimezone)	calTimezone = icsTimezone;
	if (verbose)	console.log('calTimezone: ', calTimezone);

	// page is now ready, initialize the calendar...
	var calendarEl = document.getElementById('calendar');
	var calendar = new FullCalendar.Calendar(calendarEl, {
		plugins: [ 'interaction', 'dayGrid', 'rrule', 'moment', 'momentTimezone' ],	// docs on plugin names are wrong !!
		timeZone: calTimezone,	//	setting from config, default is 'local'
		locale: LocaleCode,
		weekNumbers: weekNums,
		header: {
			left: 'prevYear,nextYear',
			center: 'title',
		},
		navLinks: false, // can click day/week names to navigate views
		editable: false,
		eventLimit: false, // allow "more" link when too many events
		fixedWeekCount: false,
        firstDay: firstWeekDay,    // new 20.03.23 - DONE: make this configurable !
		eventClick: function(info) {
			info.jsEvent.preventDefault(); // don't let the browser navigate
			//	console.log(info.event.extendedProps["description"]);
			if (info.event.url) {
				window.open(info.event.url);	// open url in new Window/Tab
			}
			if (enableDescPopup)	{	// show alert popup with description, if enabled
				if (info.event.extendedProps["description"]) {
					alert(info.event.extendedProps["description"]);
				}
			}
		},
		//	Description as Tooltip (tippy.js) :
		eventRender: function(info) {
			if (!enableDescPopup) {	// tippy hover only if click popup is disabled
				if (info.event.extendedProps.description) {
					tippy (info.el, {
						content: info.event.extendedProps.description,
						allowHTML: true,	// see https://github.com/wernerjoss/grav-plugin-fullcalendar/issues/29
					});
				}
			}
		},
		events: function(info, successCallback, failureCallback) {
			var allevents = [];
			calUrls.forEach(function(value, index) {
				calendarUrl = value;
				if (verbose) console.log('Calendar URL:' + calendarUrl);
				var events = [];
				var do_callback = false; // muss zwingend hier hin, nicht ausserhalb der forEach schleife !!
				if (index == (len - 1)) {
					do_callback = true;
				}
				if (verbose) console.log('index,do_callback:', index, do_callback);
				jQuery.ajax({
					crossOrigin: true,
					proxy: cors_api_url,	//	"http://localhost:8080/proxy.php", //to overide default proxy
					url: calendarUrl,
					//dataType: "json", //no need. if you use crossOrigin, the dataType will be override with "json"
					//charset: 'ISO-8859-1', //use it to define the charset of the target url
					//contentType: "application/json; charset=ISO-8859-1",	// test encoding, geht nicht 31.08.22
					context: {},
					success: function(data) {
						//	alert(data);
						//	$( '#test' ).html(data);
					}
				})
				.done(function( data, textStatus, jqXHR ) {
					if (verbose)	console.log(data);
					var jcalData = ICAL.parse(data);	//	directly parse data, no need to split to lines first ! 14.02.20
					var comp = new ICAL.Component(jcalData);

					var comp = new ICAL.Component(jcalData);
					var eventComps = comp.getAllSubcomponents("vevent");
					//	map them to FullCalendar events Objects
					events = jQuery.map(eventComps, function(item) {
						var fcevents = {};
						//	fcevents["tzid"] = icsTimezone;
						var entry = item.getFirstPropertyValue("summary");
						//	console.log('Entry: ', entry);
						//	fix badly encoded Text from remote google calendars 01.09.22
						try	{	// see https://stackoverflow.com/questions/5396560/how-do-i-convert-special-utf-8-chars-to-their-iso-8859-1-equivalent-using-javasc
							entry = decodeURIComponent(escape(entry));
						}	catch(e)	{
							entry = entry;
						}
						if (entry !== null)	fcevents["title"] = entry;
						var entry = item.getFirstPropertyValue("location");
						if (entry !== null)	fcevents["location"] = entry;
						var entry = item.getFirstPropertyValue("url");
						if (entry !== null)	fcevents["url"] = entry;
						var entry = item.getFirstPropertyValue("dtstart");
						if (entry !== null)	{ fcevents["start"] = entry.toJSDate(); var start = entry;}

						//	tz_offset_single = 0;	//		now from plugin config 01.01.22
						if (verbose) console.log('tz_offset_single:', tz_offset_single);
						if (tz_offset_single != 0) {
							start["hour"] = (start["hour"] + Number(tz_offset_single)) % 24;	// add hours from config, type conversion mandatory ! :)
							fcevents["start"] = start.toJSDate();
							if (verbose) console.log('newstart', start);
						}
						//	tz_offset_minutes = 15;	//
						if (verbose) console.log('tz_offset_minutes:', tz_offset_minutes);
						if (tz_offset_minutes != 0) {
							start["minute"] = (start["minute"] + Number(tz_offset_minutes));	// add minutes from config, do NOT use modulo 60 !
							fcevents["start"] = start.toJSDate();
							if (verbose) console.log('newstart', start);
						}

						var entry = item.getFirstPropertyValue("dtend");
						if (entry !== null)	{ fcevents["end"] = entry.toJSDate(); var end = entry; }
						duration = fcevents["end"] - fcevents["start"];	// calculate event duration 29.08.20
						if (verbose)	console.log('Duration:', duration);
						fcevents["allDay"] = true;	// default value -> span .fc-time in grid is NOT created
						if (duration < 86400000)	fcevents["allDay"] = false;	// duration less than 1 day: allDay = false
						var entry = item.getFirstPropertyValue("description");	// add description 22.06.20
						try	{	// see https://stackoverflow.com/questions/5396560/how-do-i-convert-special-utf-8-chars-to-their-iso-8859-1-equivalent-using-javasc
							entry = decodeURIComponent(escape(entry));
						}	catch(e)	{
							entry = entry;
						}
						if (entry !== null)	fcevents["description"] = entry;
						var entry = item.getFirstPropertyValue("color");	// add color from ics
						if (entry !== null)	fcevents["color"] = entry;

						// not used options go here

						var rrules = item.getFirstPropertyValue("rrule");
						var fcrrules = {};	// extra object for rrules
						if (rrules !== null)	{
							if (rrules.freq !== null)	{	//	freq is required, do not continue if null
								if (verbose)	console.log('rrules:', rrules);
								fcrrules["freq"] = rrules.freq;
								if (verbose) console.log('start', start["_time"]);
								if (verbose) console.log('start day', start["day"]);
								// Korrektur f. 1 Tag Versatz nach vorne 06.03.22 - crazy - see https://github.com/wernerjoss/grav-plugin-fullcalendar/issues/44#issuecomment-1057087028
								if (fcevents["allDay"]) {
									start["day"] = start["day"] + 1;
									end["day"] = end["day"] + 1;
									if (verbose) console.log('start day', start["day"], 'end day', end["day"]);
								}
								if (verbose) console.log('tz_offset_rec:', tz_offset_rec);	//	now from plugin config 01.01.22
								var tz_offset_dl = parseInt(tz_offset_rec);	// default
								M = parseInt(start["month"]);
								//	console.log('Month:', M);	//	start["month"]);
								if ((M > dlstart) && (M <= dlend)) {
									tz_offset_dl = tz_offset_dl + 1;	// add 1h during DAYLIGHT saving period
								}	// DONE: get DAYLIGHT Start/Endmonth from ICS
								//	console.log(tz_offset_dl);
								if (tz_offset_dl != 0) {
									start["hour"] = (start["hour"] + Number(tz_offset_dl)) % 24;	// add hours from config, type conversion mandatory ! :)
									//	fcevents["start"] = start.toJSDate();
									if (verbose) console.log('newstart', start);
								}
								fcevents["start"] = start.toJSDate();	// move here 06.03.22 !
								/* not needed
								end["hour"] = end["hour"] + tz_offset;
								if (verbose)	console.log('newend', end);
								fcevents["end"] = end.toJSDate();
								*/
								var parts = rrules["parts"];
								if (verbose)	console.log('parts:', parts);
								var byweekday = parts["BYDAY"];
								var weekdays = [];	// must be empty array, otherwise, push() will not work !
								var bysetpos = [];
								if (Array.isArray(byweekday))	{
									byweekday = parts["BYDAY"];
									for (i = 0; i < byweekday.length; i++) {
										//	DONE: implement BYDAY n+ or n-
										if (byweekday[i].match(/\d+/g))	{	// entry contains digits, save them to setpos, strip from weekdays
											var daynum = parseInt(byweekday[i]).toString();
											//	console.log('daynum: ' + daynum) ;
											bysetpos.push(daynum);
											weekdays.push(byweekday[i].replace(/[0-9,+,-]/g, ''));
										} else { weekdays.push(byweekday[i]); }	// no digits, just save to weekdays
									}
									byweekday = weekdays;
								}	else	{byweekday = null;}
								if (verbose)	console.log('byweekday:', byweekday);
								var byweekno = parts["BYWEEKNO"];
								if (Array.isArray(byweekno))	{byweekno = parts["BYWEEKNO"];}	else	{byweekno = null;}
								if (verbose)	console.log('byweekno:', byweekno);
								var bymonth = parts["BYMONTH"];
								if (Array.isArray(bymonth))	{bymonth = parts["BYMONTH"];}	else	{bymonth = null;}
								if (verbose)	console.log('bymonth:', bymonth);
								var bymonthday = parts["BYMONTHDAY"];
								if (Array.isArray(bymonthday))	{bymonthday = parts["BYMONTHDAY"];}	else	{bymonthday = null;}
								if (verbose)	console.log('bymonthday:', bymonthday);
								var byyearday = parts["BYYEARDAY"];
								if (Array.isArray(byyearday))	{byyearday = parts["BYYEARDAY"];}	else	{byyearday = null;}
								if (verbose)	console.log('byyearday:', byyearday);
								if (rrules.dtstart !== undefined)	{fcrrules["dtstart"] = rrules.dtstart;}	else	{fcrrules["dtstart"] = fcevents["start"];}
								if (byweekday !== null) { fcrrules["byweekday"] = byweekday;}
								if (bysetpos !== null) { fcrrules["bysetpos"] = bysetpos;}
								if (byweekno !== null) { fcrrules["byweekno"] = byweekno;}
								if (bymonth !== null) { fcrrules["bymonth"] = bymonth;}
								if (bymonthday !== null) { fcrrules["bymonthday"] = bymonthday;}
								if (byyearday !== null) { fcrrules["byyearday"] = byyearday;}
								if (rrules.interval != null) { fcrrules["interval"] = rrules.interval;}
								if (rrules.count != null) { fcrrules["count"] = rrules.count;}
								if (rrules.wkst != null) { fcrrules["wkst"] = rrules.wkst;}
								if (rrules.until != null) { fcrrules["until"] = rrules.until.toJSDate();}

								fcevents["rrule"] = fcrrules;
								if (verbose) console.log('fcrrules:', fcrrules);
							}
						}
						if(fcevents["color"] == null) { fcevents["backgroundColor"] = colors[index];}
						if (verbose)	console.log('fcevents:', fcevents);
						if (item.getFirstPropertyValue("class") === "PRIVATE") {
							return null;
						} else {
							return fcevents;
						}
					})
					jQuery.merge(allevents, events);
					if (verbose) console.log('index,do_callback:', index, do_callback);
					if (verbose) console.log('events:', events);
					if (do_callback) {
						successCallback(allevents);	// wichtig !!
						if (verbose)	console.log('allevents:', allevents);
					}
				},
				'text');
			})
		}
	});
	//	console.log(calendar);
	calendar.render();
	// show legend, if enabled
	if (showlegend) {
		// Add the contents of cfgfiles to #legend:
		document.getElementById('legend').appendChild(makeUL(calNames, colors));
	}
}

function makeUL(array, colors) {
	// Create the list element:
	var list = document.createElement('ul');
	// assign css class
	list.classList.add('cal_legend');
	for (var i = 0; i < array.length; i++) {
		// Create the list item:
		var item = document.createElement('li');

		// Set its contents:
		item.appendChild(document.createTextNode(array[i]));
		item.style.color = colors[i];

		// Add it to the list:
		list.appendChild(item);
	}
	// Finally, return the constructed list:
	return list;
}

function getAbsolutePath() { // see https://www.sitepoint.com/jquery-current-page-url/
	var loc = window.location;
	//  console.log('window.location:', loc);
	var pathName = loc.pathname.substring(0, loc.pathname.lastIndexOf('/') + 1);
	return loc.href.substring(0, loc.href.length - ((loc.pathname + loc.search + loc.hash).length - pathName.length));
}
